File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) 
    {
        vector<vector<int>> ans;
        if(!root) return ans;

        queue<TreeNode*>q;
        q.push(root);
        int count = 0;

        while(!q.empty())
        {
            vector<int>temp;
            int n = q.size();
            while(n--)
            {
                TreeNode *start = q.front();
                q.pop();
                temp.push_back(start->val);

                if(start->left) q.push(start->left);
                if(start->right) q.push(start->right);

            }
            if(count % 2 == 1) reverse(temp.begin(),temp.end());
            ans.push_back(temp);
            count++;
        }
        return ans;

    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        if not root:
            return []

        ans = []
        q = deque([root])

        left_to_right = True

        while q:
            l = len(q)
            temp = [0] * l

            for i in range(l):

                node = q.popleft()
                
                index = i if left_to_right else (l - 1 - i)
                temp[index] = node.val

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            
            
            ans.append(temp)
            left_to_right = not left_to_right

        return ans
            